### PR TITLE
cookbook_hierarchical_clustering

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

File: examples/meta/src/classifier/hierarchical.sg [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+Hierarchical hierarchical(merges, distance)
+#![create_instance]
+
+#![train_model]
+hierarchical.train()
+#![train_model]
+
+#![extract_results]
+RealVector d = hierarchical.get_merge_distances()
+RealMatrix cp = hierarchical.get_cluster_pairs()
```
Comment (karlnapf) [inline examples/meta/src/classifier/hierarchical.sg]: Check the method signature:
`SGMatrix<int32_t> get_cluster_pairs();`
This make sthe csharp fail on travis (strongly typed)

File: doc/cookbook/source/examples/classifier/hierarchical.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
+=======================
+Hierarchical Clustering
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: Should this be in "clustering" or in "classifier"?

File: doc/cookbook/source/examples/classifier/hierarchical.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+=======================
+
+Hierarchical clustering is a method of cluster analysis which seeks to build a hierarchy of clusters.
+In Shogun, we applies "bottom up" approach for the hierarchical clustering: each observation starts in its own cluster, and pairs of clusters are merged as one moves up the hierarchy.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: We apply a "bottom up" approach: each observation starts in its own clister, and pairs of clusters are subsequently merged.

Cleaner language

File: doc/cookbook/source/examples/classifier/hierarchical.rst [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+The merges are determined in a greedy manner. 
+To start with, we construct a distance matrix, where it saves the distance between the i-th and j-th elements.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: watch your English :)
Use something as "We start by constructing a pairwise distance matrix. Then, the clusters of the pair with closest distance are merged iteratively."

File: doc/cookbook/source/examples/classifier/hierarchical.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+Imagine we have files with the training data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) as:
+
+.. sgexample:: hierarchical.sg:create_features
+
+In order to run :sgclass:`CHierarchical`, we need to choose a distance, for example :sgclass:`CEuclideanDistance`, or other sub-classes of :sgclass:`CDistance`. The distance is initialized with the data we want to classify.
+
+.. sgexample:: hierarchical.sg:choose_distance
+
+We then create an instance of the :sgclass:`CHierarchical` classifier by passing it `merges` and `CDistance`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: use sgclass for CDistance.
Never mention method names in the text please, let the code speak

File: doc/cookbook/source/examples/classifier/hierarchical.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+`Merges` assigns the steps of merging we expect to have in the training. 
+
+.. sgexample:: hierarchical.sg:create_instance
+
+And we can extract the information of the two merged elements, as while as the distance between them in each merging step:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: Remove "And"
"while" -> well

File: doc/cookbook/source/examples/classifier/hierarchical.rst [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+.. sgexample:: hierarchical.sg:extract_results
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: if you put this example in "classifier" please also show how to extract predicted labels

**Comment by karlnapf** [comment]
Thanks for the patch!

 * see my comments
 * add integration testing data (see readme of cookbook, separate patch, need to update data version here)
 * squash commits into one
 * done :)

File: doc/cookbook/source/examples/classifier/hierarchical.rst (hunk 2 above)
Comment (OXPHOS) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: Currently all clustering methods are under classifier entry..but it should be hierarchical

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/hierarchical.rst]: yeah put them under clustering then, there are no labels given ....

**Comment by karlnapf** [comment]
Please ping me next time you update, I did not see :) (not notified when pr is udpates, only when there is a message below)

**Comment by karlnapf** [comment]
ah...sorry, the integration testing data is missing, you need to update the data version to the latest

**Comment by OXPHOS** [comment]
@karlnapf  sry can you explain more about the data version?
I tried `git commit data` in shogun, this branch, and `git status` shows:
```
On branch cookbook_hierarchical_clustering
Your branch is up-to-date with 'origin/cookbook_hierarchical_clustering'.
nothing to commit, working directory clean
```

**Comment by karlnapf** [comment]
You sent this data patch:
https://github.com/shogun-toolbox/shogun-data/pull/87

It has commit hash: c2f6abd1a57e361a5222592334c9a15b6f02e8dc
The current data hash in the main repository (which tells shogun which revision of the data repo to use) is 1054d4f6a0a305540ddfdbbe6f4c05e14184efd1 
As you see, [here](https://github.com/shogun-toolbox/shogun-data/tree/1054d4f6a0a305540ddfdbbe6f4c05e14184efd1), the data revision is an earlier one than your patch. This means that the integration tests (that the meta examples are part of) can't test this particular example.

In order to update, do `git submodule update` or pull it manually. After having done this, the `data` revision in the main repo should contain the latest hash of the data repo and you can add that to your PR here.

Let me know if you still have questions on it, and don't feel bad about them ... takes a while to get through all that ;)

**Comment by OXPHOS** [comment]
@karlnapf  
I did `git submodule add https://github.com/shogun-toolbox/shogun-data.git` and then `git submodule update`. I guess this is not correct..?

**Comment by karlnapf** [comment]
`git submodule intit` should also do it.
But it is just a clone of a repository, you can cd to `shogun/data` and do `git pull` as well

File: shogun-data [hunk 8]
```diff
@@ -0,0 +1 @@
+Subproject commit 9b12ffc6add78290c135025f913fab488657ff5d
```
Comment (karlnapf) [inline shogun-data]: yes, that is it! :)

Comment (karlnapf) [inline shogun-data]: Ah, not 100%.
This hash points to a PR by somebody else that is more recent than yours.
Your hash was a different one. You can either revert the data repo to your hash and then update the subproject again, or manually put your hash in.

It doesnt really matter here as we will merge all the data and cookbooks anyways, but you need to make sure that you are in controll of this

Comment (karlnapf) [inline shogun-data]: This one doesn't work -- see output of travis

Comment (karlnapf) [inline shogun-data]: OK, I just merged #3219 which includes your hierarchical data.
This means you can remove the data version at all, and we can merge.

 However, make sure to understand this to have an easier life for the next cookbook ;)

**Comment by OXPHOS** [comment]
@karlnapf this one is ready I think.

```
Start 266: integration_meta_cpp-clustering-hierarchical
271/387 Test #266: integration_meta_cpp-clustering-hierarchical ..........................   Passed    0.02 sec
``

**Comment by karlnapf** [comment]
Thanks!

**Pull request merged**